backend/log/context_grpc.go

```diff
@@ -0,0 +1,55 @@
+package log
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"google.golang.org/grpc/metadata"
+)
+
+const (
+	logParamsCtxMetadataKey = "loggerParamsCtxMetadata"
+	logParamSeparator       = ":"
+)
+
+// WithContextualAttributesForOutgoingContext will append the given key/value log parameters to the outgoing context.
+func WithContextualAttributesForOutgoingContext(ctx context.Context, logParams []any) context.Context {
+	if len(logParams) == 0 || len(logParams)%2 != 0 {
+		return ctx
+	}
+
+	for i := 0; i < len(logParams); i += 2 {
+		k := logParams[i].(string)
+		v := logParams[i+1].(string)
+		if k == "" || v == "" {
+			continue
+		}
+
+		ctx = metadata.AppendToOutgoingContext(ctx, logParamsCtxMetadataKey, logParam(k, v))
+	}
+
+	return ctx
+}
+
+// ContextualAttributesFromIncomingContext returns the contextual key/value log parameters from the given incoming context.
+func ContextualAttributesFromIncomingContext(ctx context.Context) []any {
+	logParams := metadata.ValueFromIncomingContext(ctx, logParamsCtxMetadataKey)
+	if len(logParams) == 0 {
+		return nil
+	}
+
+	var attrs []any
+	for _, param := range logParams {
+		kv := strings.Split(param, logParamSeparator)
+		if len(kv) != 2 || kv[0] == "" || kv[1] == "" {
```

Comment:
an empty value is not a valid value?

Reply:
Yeah I wanted to keep things simple for now. I don't foresee this being used heavily as it's quite a special use case. Can always revisit if it becomes important

Reply:
I think it would be useful to see if a value is set but it's empty but maybe I am missing something so up to you.